spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java

```diff
@@ -130,15 +131,22 @@ public brave.Tracer braveTracer(Tracing tracing) {
 	}
 
 	@Bean
+	@Scope("prototype")
 	@ConditionalOnMissingBean
-	public CurrentTraceContext braveCurrentTraceContext(List<CurrentTraceContext.ScopeDecorator> scopeDecorators,
+	public CurrentTraceContext.Builder braveCurrentTraceContextBuilder() {
+		return ThreadLocalCurrentTraceContext.newBuilder();
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public CurrentTraceContext braveCurrentTraceContext(CurrentTraceContext.Builder currentTraceContextBuilder,
+			List<CurrentTraceContext.ScopeDecorator> scopeDecorators,
 			List<CurrentTraceContextCustomizer> currentTraceContextCustomizers) {
-		ThreadLocalCurrentTraceContext.Builder builder = ThreadLocalCurrentTraceContext.newBuilder();
-		scopeDecorators.forEach(builder::addScopeDecorator);
+		scopeDecorators.forEach(currentTraceContextBuilder::addScopeDecorator);
 		for (CurrentTraceContextCustomizer currentTraceContextCustomizer : currentTraceContextCustomizers) {
-			currentTraceContextCustomizer.customize(builder);
+			currentTraceContextCustomizer.customize(currentTraceContextBuilder);
```

Comment:
I think that any `CurrentTraceContextCustomizer` beans should be applied to the `braveCurrentTraceContextBuilder` bean rather than only applying them when creating the `CurrentTraceContext`. This will align with other areas where we define a builder bean such as `Jackson2ObjectMapperBuilder` and `RestTemplateBuilder`.

Reply:
Is this what you intend to do?

```java
@Bean
@Scope("prototype")
@ConditionalOnMissingBean
public CurrentTraceContext.Builder braveCurrentTraceContextBuilder(List<CurrentTraceContextCustomizer> currentTraceContextCustomizers) {
	CurrentTraceContext.Builder builder =ThreadLocalCurrentTraceContext.newBuilder();
	for (CurrentTraceContextCustomizer currentTraceContextCustomizer : currentTraceContextCustomizers) {
		currentTraceContextCustomizer.customize(builder);
	}
	return builder;
}
```

But if we do, we will have to write code to iterate and apply the CurrentTraceContextCustomizers each time you want to change the type of CurrentTraceContext. I think this is inconvenient.

```java
@Bean
@Scope("prototype")
public CurrentTraceContext.Builder braveCurrentTraceContextBuilder(List<CurrentTraceContextCustomizer> currentTraceContextCustomizers) {
	CurrentTraceContext.Builder builder = StrictCurrentTraceContext.newBuilder();
	for (CurrentTraceContextCustomizer currentTraceContextCustomizer : currentTraceContextCustomizers) {
		currentTraceContextCustomizer.customize(builder);
	}
	return builder;
}
```

Reply:
Yes, for consistency with how other builders and applying customisers to them, I think we need to do something like this.

Taking a step back, perhaps we would be better offering a configuration property that allows control of the type of builder that is auto-configured. Its value could be thread local, strict, or standard which would map to `ThreadLocalCurrentTraceContext.Builder`, `StrictCurrentTraceContext.Builder` and `CurrentTraceContext.Builder` respectively.

Reply:
Hmmm. How about putting `Supplier<CurrentTraceContext.Builder>` in the bean then?
This would make it easy to change the type of CurrentTraceContext.

```java
@Bean
@ConditionalOnMissingBean
public Supplier<CurrentTraceContext.Builder> braveCurrentTraceContextBuilderSupplier() {
	return ThreadLocalCurrentTraceContext::newBuilder;
}

@Bean
@Scope("prototype")
@ConditionalOnMissingBean
public CurrentTraceContext.Builder braveCurrentTraceContextBuilder(
		Supplier<CurrentTraceContext.Builder> builderSupplier,
		List<CurrentTraceContext.ScopeDecorator> scopeDecorators,
		List<CurrentTraceContextCustomizer> currentTraceContextCustomizers) {
	CurrentTraceContext.Builder builder = builderSupplier.get();
	scopeDecorators.forEach(builder::addScopeDecorator);
	for (CurrentTraceContextCustomizer currentTraceContextCustomizer : currentTraceContextCustomizers) {
		currentTraceContextCustomizer.customize(builder);
	}
	return builder;
}

@Bean
@ConditionalOnMissingBean
public CurrentTraceContext braveCurrentTraceContext(CurrentTraceContext.Builder currentTraceContextBuilder) {
	return currentTraceContextBuilder.build();
}
```

Reply:
Sorry, I don't think that's a good idea. I can't think of anywhere else where we have the indirection of a `Supplier` providing access to a builder. I don't think there's anything that's unique about this situation to justify it and it's likely to cause confusion.

Reply:
I think it is an idea that would make it easier to use different CurrentTraceContexts, but yes, Supplier may be confused.

Hmmm, ok, I'll close this PR, thanks.
Workaround by registering my own bean.

```kotlin
@Bean
fun braveCurrentTraceContext(
    scopeDecorators: List<CurrentTraceContext.ScopeDecorator>,
    currentTraceContextCustomizers: List<CurrentTraceContextCustomizer>,
): CurrentTraceContext {
    // See: https://armeria.dev/docs/advanced-zipkin
    val builder = RequestContextCurrentTraceContext.builder()
    scopeDecorators.forEach(Consumer { scopeDecorator: CurrentTraceContext.ScopeDecorator ->
        builder.addScopeDecorator(
            scopeDecorator
        )
    })
    for (currentTraceContextCustomizer in currentTraceContextCustomizers) {
        currentTraceContextCustomizer.customize(builder)
    }
    return builder.build()
}
```